jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiRequest.java

```diff
@@ -734,8 +737,68 @@ public Part getPart(String name) throws IOException, ServletException
     @Override
     public <T extends HttpUpgradeHandler> T upgrade(Class<T> handlerClass) throws IOException, ServletException
     {
-        // Not implemented. Throw ServletException as per spec.
-        throw new ServletException("Not implemented");
+        Response response = _servletContextRequest.getServletContextResponse();
+        if (response.getStatus() != HttpStatus.SWITCHING_PROTOCOLS_101)
+            throw new IllegalStateException("Response status should be 101");
+        if (response.getHeaders().get("Upgrade") == null)
+            throw new IllegalStateException("Missing Upgrade header");
+        if (!"Upgrade".equalsIgnoreCase(response.getHeaders().get("Connection")))
+            throw new IllegalStateException("Invalid Connection header");
+        if (response.isCommitted())
+            throw new IllegalStateException("Cannot upgrade committed response");
+        if (_servletChannel.getConnectionMetaData().getHttpVersion() != HttpVersion.HTTP_1_1)
+            throw new IllegalStateException("Only requests over HTTP/1.1 can be upgraded");
+
+        ServletOutputStream outputStream = _servletContextRequest.getHttpOutput();
+        ServletInputStream inputStream = _servletContextRequest.getHttpInput();
```

Comment:
Isn't is sufficient just to wrap the input and output streams to check for completeness:

- if close() is called on either of them
- if a read returns -1 or throws an exception
- if a write throws and exception.

This doesn't need to be a perfect mechanism. We could also add some idle timeout mechanism as a catch all.